# Лабораторная работа №6. Random Forest

Классификация типа поверхности по сигналам IMU-датчиков робота (соревнование
Kaggle **CareerCon 2019**). Лес решающих деревьев реализован вручную; служебные
операции (критерий энтропии, метрики, ROC) вынесены в библиотеку проекта
`mlcore`.

**Что требуется по ТЗ:** дерево внутри леса — не бинарное и не сбалансированное;
ограничение глубины — обосновать; посчитать accuracy / precision / recall / F1;
построить ROC-кривую.

## 1. Краткая теория

Решающее дерево разбивает выборку $S$ в каждом узле так, чтобы максимизировать
прирост информации
$$IG = H(S) - \sum_i \frac{|S_i|}{|S|} H(S_i), \qquad
H(S) = -\sum_c p_c \log_2 p_c .$$

Чтобы узлы были **multiway** (не бинарные), признак режется на $m \in \{2,3,4\}$
квантильных диапазона, и среди всех $(f, m)$ выбирается вариант с наибольшим $IG$.
Ветви растут независимо и упираются в разные критерии остановки, поэтому дерево
получается **не сбалансированным**.

`Random Forest` усредняет вероятности классов $T$ деревьев, обученных на
bootstrap-выборках; в каждом узле рассматривается случайные $\sqrt{d}$ признаков.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from mlcore.utils.notebook import lab_paths, setup_plotting
from mlcore.utils.timing import timer
from mlcore.preprocessing.encoding import label_encode
from mlcore.preprocessing.splitting import train_test_split
from mlcore.trees import entropy
from mlcore.metrics import (
    accuracy, confusion_matrix, precision, recall, f1_score,
    roc_curve, auc_score,
)

ROOT, _, ASSETS = lab_paths(6)
setup_plotting()
DATA = ROOT / 'labs/lr-6/data'
np.random.seed(42)

## 2. Данные и разведочный анализ

Одна серия (`series_id`) — это проход робота длиной 128 замеров по 10 каналам
(кватернион ориентации $X,Y,Z,W$, угловая скорость и линейное ускорение по трём
осям). Метка `surface` задана на уровне серии.

In [2]:
X_raw = pd.read_csv(DATA / 'X_train.csv')
y_raw = pd.read_csv(DATA / 'y_train.csv')

series_len = X_raw.groupby('series_id').size().unique()
print(f'X_raw {X_raw.shape}, y_raw {y_raw.shape}')
print(f'Замеров на серию: {series_len}')
print(f'Серий {y_raw.series_id.nunique()}, классов {y_raw.surface.nunique()}')

X_raw (487680, 13), y_raw (3810, 3)
Замеров на серию: [128]
Серий 3810, классов 9


In [3]:
counts = y_raw.surface.value_counts()

fig, ax = plt.subplots(figsize=(8, 4))
ax.bar(counts.index, counts.values, color='steelblue', edgecolor='black')
ax.set_title('Распределение классов поверхностей')
ax.set_ylabel('Количество серий')
ax.grid(True, alpha=0.3, axis='y')
plt.xticks(rotation=45, ha='right')
fig.tight_layout()
fig.savefig(ASSETS / 'class_distribution.png', dpi=150)
plt.close(fig)
print(counts.to_string())

surface
concrete                  779
soft_pvc                  732
wood                      607
tiled                     514
fine_concrete             363
hard_tiles_large_space    308
soft_tiles                297
carpet                    189
hard_tiles                 21


Классы сильно несбалансированы (от 779 серий `concrete` до 21 `hard_tiles`),
поэтому одной accuracy недостаточно — ниже считаем precision/recall/F1 по каждому
классу и в усреднениях macro / weighted.

## 3. Признаки

Временной ряд каждого канала сворачивается в 7 статистик (среднее, std, min, max,
медиана, размах и mean-abs-change $\tfrac{1}{T-1}\sum_t|s_{t+1}-s_t|$). Добавляем
два производных канала — модули угловой скорости и ускорения. Итого
$12 \times 7 = 84$ признака на серию.

In [4]:
RAW_CHANNELS = [
    'orientation_X', 'orientation_Y', 'orientation_Z', 'orientation_W',
    'angular_velocity_X', 'angular_velocity_Y', 'angular_velocity_Z',
    'linear_acceleration_X', 'linear_acceleration_Y', 'linear_acceleration_Z',
]

STATS = {
    'mean':  lambda g: g.mean(),
    'std':   lambda g: g.std(),
    'min':   lambda g: g.min(),
    'max':   lambda g: g.max(),
    'med':   lambda g: g.median(),
    'range': lambda g: g.max() - g.min(),
    'mac':   lambda g: g.apply(lambda s: np.mean(np.abs(np.diff(s.values)))),
}

def make_features(raw):
    raw = raw.copy()
    raw['ang_vel_mag'] = np.sqrt(
        raw.angular_velocity_X ** 2 + raw.angular_velocity_Y ** 2 + raw.angular_velocity_Z ** 2
    )
    raw['lin_acc_mag'] = np.sqrt(
        raw.linear_acceleration_X ** 2 + raw.linear_acceleration_Y ** 2 + raw.linear_acceleration_Z ** 2
    )
    grouped = raw.groupby('series_id')
    columns = {}
    for channel in RAW_CHANNELS + ['ang_vel_mag', 'lin_acc_mag']:
        col = grouped[channel]
        for name, fn in STATS.items():
            columns[f'{channel}_{name}'] = fn(col)
    return pd.DataFrame(columns)

with timer('feature engineering'):
    features = make_features(X_raw)
features = features.reindex(y_raw.series_id.values)
print(f'Матрица признаков {features.shape}')

[timer] (feature engineering): 0.4454s
Матрица признаков (3810, 84)


In [5]:
y, label_to_idx = label_encode(y_raw.surface)
labels = sorted(label_to_idx, key=label_to_idx.get)
idx_to_label = {i: name for name, i in label_to_idx.items()}

X = np.nan_to_num(features.values.astype(np.float64))
feature_names = list(features.columns)
N_CLASSES = len(labels)
print(f'X {X.shape}, классов {N_CLASSES}')
print(labels)

X (3810, 84), классов 9
['carpet', 'concrete', 'fine_concrete', 'hard_tiles', 'hard_tiles_large_space', 'soft_pvc', 'soft_tiles', 'tiled', 'wood']


## 4. Разбиение train / test

Разметка есть только у train-части Kaggle, поэтому делим её сами —
стратифицированно по классам в пропорции 75 / 25.

In [6]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=42, stratify=y,
)
print(f'Train {X_train.shape}, Test {X_test.shape}')
for k in range(N_CLASSES):
    n_tr = int((y_train == k).sum())
    n_te = int((y_test == k).sum())
    print(f'  {idx_to_label[k]:25s} train={n_tr:4d}  test={n_te:4d}')

Train (2858, 84), Test (952, 84)
  carpet                    train= 142  test=  47
  concrete                  train= 584  test= 195
  fine_concrete             train= 272  test=  91
  hard_tiles                train=  16  test=   5
  hard_tiles_large_space    train= 231  test=  77
  soft_pvc                  train= 549  test= 183
  soft_tiles                train= 223  test=  74
  tiled                     train= 386  test= 128
  wood                      train= 455  test= 152


## 5. Решающее дерево (multiway)

Узел перебирает случайные $\sqrt{d}$ признаков, для каждого пробует разбиение на
2–4 квантильных диапазона и фиксирует вариант с максимальным $IG$. Дочерние узлы
хранятся списком — их может быть от 2 до 4. Рост ветви останавливается по глубине,
размеру выборки, чистоте узла или при $IG \le 0$.

In [7]:
class _TreeNode:
    __slots__ = ('probs', 'pred', 'feature', 'thresholds', 'children')

    def __init__(self, probs, pred):
        self.probs = probs
        self.pred = pred
        self.feature = None
        self.thresholds = None
        self.children = None


class DecisionTree:
    def __init__(self, max_depth, min_samples, max_bins, n_features, rng, n_classes):
        self.max_depth = max_depth
        self.min_samples = min_samples
        self.max_bins = max_bins
        self.n_features = n_features
        self.rng = rng
        self.n_classes = n_classes
        self.root = None

    def fit(self, X, y):
        self.root = self._grow(X, y, 0)
        return self

    def _leaf(self, counts):
        return _TreeNode(counts / counts.sum(), int(np.argmax(counts)))

    def _best_split(self, X, y, parent_h):
        candidates = self.rng.choice(
            X.shape[1], size=min(self.n_features, X.shape[1]), replace=False
        )
        best_gain, best_feature, best_thr = 0.0, None, None
        for f in candidates:
            column = X[:, f]
            if np.unique(column).size < 2:
                continue
            for m in range(2, self.max_bins + 1):
                edges = np.quantile(column, np.linspace(0, 1, m + 1))
                thr = np.unique(edges[1:-1])
                if thr.size < 1:
                    continue
                buckets = np.digitize(column, thr)
                gain = parent_h
                for b in np.unique(buckets):
                    gain -= (np.count_nonzero(buckets == b) / len(y)) * entropy(y[buckets == b])
                if gain > best_gain:
                    best_gain, best_feature, best_thr = gain, f, thr
        return best_gain, best_feature, best_thr

    def _grow(self, X, y, depth):
        counts = np.bincount(y, minlength=self.n_classes).astype(float)
        node = self._leaf(counts)
        if depth >= self.max_depth or len(y) < self.min_samples or np.count_nonzero(counts) == 1:
            return node

        gain, feature, thr = self._best_split(X, y, entropy(y))
        if feature is None or gain <= 1e-9:
            return node

        node.feature, node.thresholds, node.children = feature, thr, []
        buckets = np.digitize(X[:, feature], thr)
        for b in range(thr.size + 1):
            mask = buckets == b
            node.children.append(
                self._leaf(counts) if not mask.any() else self._grow(X[mask], y[mask], depth + 1)
            )
        return node

    def _route(self, x):
        node = self.root
        while node.children is not None:
            node = node.children[int(np.digitize(x[node.feature], node.thresholds))]
        return node

    def predict_proba(self, X):
        out = np.empty((len(X), self.n_classes))
        for i in range(len(X)):
            out[i] = self._route(X[i]).probs
        return out

    def leaf_depths(self):
        acc = []
        def walk(node, depth):
            if node.children is None:
                acc.append(depth)
            else:
                for child in node.children:
                    walk(child, depth + 1)
        walk(self.root, 0)
        return acc

    def branching(self):
        hist = {}
        def walk(node):
            if node.children is None:
                return
            hist[len(node.children)] = hist.get(len(node.children), 0) + 1
            for child in node.children:
                walk(child)
        walk(self.root)
        return hist

## 6. Random Forest

Бэггинг: каждое дерево учится на bootstrap-выборке, рандомизация признаков идёт
из общего генератора `rng`. Предсказание — усреднение вероятностей деревьев и
`argmax`.

In [8]:
class RandomForest:
    def __init__(self, n_trees=60, max_depth=12, min_samples=5, max_bins=4,
                 random_state=42, n_classes=None):
        self.n_trees = n_trees
        self.max_depth = max_depth
        self.min_samples = min_samples
        self.max_bins = max_bins
        self.random_state = random_state
        self.n_classes = n_classes
        self.trees = []

    def fit(self, X, y):
        if self.n_classes is None:
            self.n_classes = len(np.unique(y))
        rng = np.random.RandomState(self.random_state)
        n_features = max(1, int(np.sqrt(X.shape[1])))
        n = len(y)
        self.trees = []
        for _ in range(self.n_trees):
            sample = rng.randint(0, n, n)
            tree = DecisionTree(self.max_depth, self.min_samples, self.max_bins,
                                n_features, rng, self.n_classes)
            self.trees.append(tree.fit(X[sample], y[sample]))
        return self

    def predict_proba(self, X):
        total = np.zeros((len(X), self.n_classes))
        for tree in self.trees:
            total += tree.predict_proba(X)
        return total / len(self.trees)

    def predict(self, X):
        return np.argmax(self.predict_proba(X), axis=1)

## 7. Обоснование `max_depth`

Прогоняем лес по сетке глубин и сравниваем train/test accuracy. Берём наименьшую
глубину, на которой test-кривая выходит на плато, — компромисс смещения и разброса.

In [9]:
depth_grid = [2, 4, 6, 8, 10, 12, 15, 20]
train_curve, test_curve = [], []
for d in depth_grid:
    forest = RandomForest(n_trees=25, max_depth=d, min_samples=5, max_bins=4,
                          random_state=1, n_classes=N_CLASSES).fit(X_train, y_train)
    train_curve.append(float((forest.predict(X_train) == y_train).mean()))
    test_curve.append(float((forest.predict(X_test) == y_test).mean()))
    print(f'max_depth={d:2d}  train={train_curve[-1]:.4f}  test={test_curve[-1]:.4f}')

best_depth = depth_grid[int(np.argmax(test_curve))]

fig, ax = plt.subplots(figsize=(8, 5))
ax.plot(depth_grid, train_curve, 'o-', label='train accuracy', linewidth=2)
ax.plot(depth_grid, test_curve, 's-', label='test accuracy', linewidth=2)
ax.axvline(best_depth, color='red', ls='--', alpha=0.7, label=f'выбор max_depth = {best_depth}')
ax.set_xlabel('Максимальная глубина дерева')
ax.set_ylabel('Accuracy')
ax.set_title('Подбор ограничения глубины')
ax.legend()
ax.grid(alpha=0.3)
fig.tight_layout()
fig.savefig(ASSETS / 'depth_justification.png', dpi=150)
plt.close(fig)
print(f'Выбранная глубина: {best_depth}')

max_depth= 2  train=0.5658  test=0.5956
max_depth= 4  train=0.8698  test=0.8015
max_depth= 6  train=0.9955  test=0.8319
max_depth= 8  train=0.9955  test=0.8319
max_depth=10  train=0.9955  test=0.8319
max_depth=12  train=0.9955  test=0.8319
max_depth=15  train=0.9955  test=0.8319
max_depth=20  train=0.9955  test=0.8319
Выбранная глубина: 6


## 8. Финальная модель

Обучаем лес из 60 деревьев с выбранной глубиной и проверяем, что узлы
действительно multiway, а ветви разной длины.

In [10]:
with timer('random forest (final)'):
    forest = RandomForest(n_trees=60, max_depth=best_depth, min_samples=5, max_bins=4,
                          random_state=1, n_classes=N_CLASSES).fit(X_train, y_train)

proba = forest.predict_proba(X_test)
pred = np.argmax(proba, axis=1)
print(f'n_trees=60, max_depth={best_depth}')
print(f'Test accuracy = {(pred == y_test).mean():.4f}')

[timer] (random forest (final)): 19.4501s
n_trees=60, max_depth=6
Test accuracy = 0.8435


In [11]:
branch_hist = {}
for tree in forest.trees:
    for k, v in tree.branching().items():
        branch_hist[k] = branch_hist.get(k, 0) + v
total_nodes = sum(branch_hist.values())
print(f'Внутренних узлов в лесу: {total_nodes}')
for k in sorted(branch_hist):
    print(f'  {k} потомков: {branch_hist[k]} ({100 * branch_hist[k] / total_nodes:.1f}%)')

all_depths = [d for tree in forest.trees for d in tree.leaf_depths()]
print(f'Глубины листьев: min={min(all_depths)}, max={max(all_depths)}, среднее={np.mean(all_depths):.2f}')

fig, ax = plt.subplots(figsize=(8, 4))
ax.hist(all_depths, bins=range(0, max(all_depths) + 2), color='steelblue', edgecolor='black')
ax.set_xlabel('Глубина листа')
ax.set_ylabel('Количество листьев (по всем деревьям)')
ax.set_title(f'Распределение глубин листьев — min={min(all_depths)}, max={max(all_depths)}')
ax.grid(True, alpha=0.3)
fig.tight_layout()
fig.savefig(ASSETS / 'leaf_depth_histogram.png', dpi=150)
plt.close(fig)

Внутренних узлов в лесу: 15943
  2 потомков: 698 (4.4%)
  3 потомков: 3303 (20.7%)
  4 потомков: 11942 (74.9%)
Глубины листьев: min=3, max=6, среднее=4.97


## 9. Метрики качества

precision/recall/F1 считаем по каждому классу (One-vs-Rest) и усредняем —
функции взяты из `mlcore.metrics`.

In [12]:
class_idx = list(range(N_CLASSES))
cm = confusion_matrix(y_test, pred, labels=class_idx)
support = cm.sum(axis=1)

with np.errstate(divide='ignore', invalid='ignore'):
    report = pd.DataFrame({
        'precision': precision(y_test, pred, average=None),
        'recall':    recall(y_test, pred, average=None),
        'f1':        f1_score(y_test, pred, average=None),
        'support':   support,
    }, index=labels)
    macro = [fn(y_test, pred) for fn in (precision, recall, f1_score)]
    weighted = [fn(y_test, pred, average='weighted') for fn in (precision, recall, f1_score)]

print(f'Accuracy = {accuracy(y_test, pred):.4f}')
print('Macro:    precision={:.4f}  recall={:.4f}  f1={:.4f}'.format(*macro))
print('Weighted: precision={:.4f}  recall={:.4f}  f1={:.4f}'.format(*weighted))
print()
print(report.round(3).to_string())

fig, ax = plt.subplots(figsize=(8, 7))
im = ax.imshow(cm, cmap='Blues')
ax.set_xticks(class_idx)
ax.set_yticks(class_idx)
ax.set_xticklabels(labels, rotation=45, ha='right')
ax.set_yticklabels(labels)
ax.set_xlabel('Предсказанный класс')
ax.set_ylabel('Истинный класс')
ax.set_title('Матрица ошибок')
for i in class_idx:
    for j in class_idx:
        ax.text(j, i, cm[i, j], ha='center', va='center',
                color='white' if cm[i, j] > cm.max() / 2 else 'black')
fig.colorbar(im, fraction=0.046, pad=0.04)
fig.tight_layout()
fig.savefig(ASSETS / 'confusion_matrix.png', dpi=150)
plt.close(fig)

Accuracy = 0.8435
Macro:    precision=0.7545  recall=0.7366  f1=0.7414
Weighted: precision=0.8416  recall=0.8435  f1=0.8397

                        precision  recall     f1  support
carpet                      0.861   0.660  0.747       47
concrete                    0.847   0.877  0.861      195
fine_concrete               0.857   0.659  0.745       91
hard_tiles                  0.000   0.000  0.000        5
hard_tiles_large_space      0.958   0.896  0.926       77
soft_pvc                    0.879   0.913  0.895      183
soft_tiles                  0.767   0.932  0.841       74
tiled                       0.843   0.883  0.863      128
wood                        0.778   0.809  0.794      152


## 10. ROC-кривые (One-vs-Rest)

Для каждого класса строим бинарную ROC по вероятностям из `predict_proba`,
считаем AUC трапециями и добавляем macro-average, усредняя TPR по общей сетке FPR.

In [13]:
grid = np.linspace(0, 1, 200)
interp = []
class_auc = {}

fig, ax = plt.subplots(figsize=(9, 7))
for k in range(N_CLASSES):
    target = (y_test == k).astype(int)
    if target.sum() == 0:
        continue
    fpr, tpr, _ = roc_curve(target, proba[:, k])
    class_auc[labels[k]] = auc_score(fpr, tpr)
    ax.plot(fpr, tpr, lw=1.5, alpha=0.85, label=f'{labels[k]} (AUC={class_auc[labels[k]]:.3f})')
    interp.append(np.interp(grid, fpr, tpr))

macro_tpr = np.mean(interp, axis=0)
macro_auc = auc_score(grid, macro_tpr)
ax.plot(grid, macro_tpr, 'k--', lw=2.5, label=f'macro-average (AUC={macro_auc:.3f})')
ax.plot([0, 1], [0, 1], color='grey', ls=':', label='случайный классификатор')
ax.set_xlabel('False Positive Rate')
ax.set_ylabel('True Positive Rate')
ax.set_title('ROC-кривые (One-vs-Rest)')
ax.legend(loc='lower right', fontsize=9)
ax.grid(alpha=0.3)
fig.tight_layout()
fig.savefig(ASSETS / 'roc_multiclass.png', dpi=150)
plt.close(fig)

print(f'macro-average AUC = {macro_auc:.4f}')
for name, val in class_auc.items():
    print(f'  {name:25s} {val:.4f}')

macro-average AUC = 0.9843
  carpet                    0.9922
  concrete                  0.9771
  fine_concrete             0.9541
  hard_tiles                0.9975
  hard_tiles_large_space    0.9949
  soft_pvc                  0.9893
  soft_tiles                0.9872
  tiled                     0.9858
  wood                      0.9799


## 11. Выводы

1. Random Forest реализован вручную: узлы режутся на 2–4 квантильных диапазона
   (большинство — на 4 ветви) — дерево **не бинарное**; глубины листьев
   различаются — **не сбалансированное**.
2. Ограничение глубины подобрано по кривой train/test accuracy — взята точка
   выхода test на плато.
3. Посчитаны accuracy, precision, recall, F1 (по классам, macro и weighted),
   построена матрица ошибок.
4. Построены ROC-кривые One-vs-Rest с AUC по классам и macro-average.